src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,167 @@
+        if (jsonTotpDevices == null) {
+            return null;
+        }
+
+        List<TotpDevice> totpDevices = new ArrayList<>();
+        for (JsonElement jsonTotpDeviceEl : jsonTotpDevices) {
+            JsonObject jsonTotpDevice = jsonTotpDeviceEl.getAsJsonObject();
+
+            String secretKey = parseAndValidateField(jsonTotpDevice, "secretKey", ValueType.STRING, true, String.class, errors, " for a totp device.");
+            Number period = parseAndValidateField(jsonTotpDevice, "period", ValueType.NUMBER, true, Number.class, errors, " for a totp device.");
+            Number skew = parseAndValidateField(jsonTotpDevice, "skew", ValueType.NUMBER, true, Number.class, errors, " for a totp device.");
+            String deviceName = parseAndValidateField(jsonTotpDevice, "deviceName", ValueType.STRING, false, String.class, errors, " for a totp device.");
+
+            if (period != null && period.intValue() < 1) {
+                errors.add("period should be > 0 for a totp device.");
+            }
+            if (skew != null && skew.intValue() < 0) {
+                errors.add("skew should be >= 0 for a totp device.");
+            }
+
+            if(deviceName != null) {
+                deviceName = deviceName.trim();
+            }
+            totpDevices.add(new TotpDevice(secretKey, period.intValue(), skew.intValue(), deviceName));
+        }
+        return totpDevices;
+    }
+
+    private static List<LoginMethod> getParsedLoginMethods(JsonObject userData, List<String> errors) {
+      JsonArray jsonLoginMethods = parseAndValidateField(userData, "loginMethods", ValueType.ARRAY_OF_OBJECT, true, JsonArray.class, errors, ".");
+
+      if (jsonLoginMethods == null) {
+          return new ArrayList<>();
+      }
+
+      if (jsonLoginMethods.size() == 0) {
+          errors.add("At least one loginMethod is required.");
+          return new ArrayList<>();
+      }
+
+      Boolean hasPrimaryLoginMethod = false;
+
+      List<LoginMethod> loginMethods = new ArrayList<>();
+      for (JsonElement jsonLoginMethod : jsonLoginMethods) {
+          JsonObject jsonLoginMethodObj = jsonLoginMethod.getAsJsonObject();
+
+          if (validateJsonFieldType(jsonLoginMethodObj, "isPrimary", ValueType.BOOLEAN)) {
+              if (jsonLoginMethodObj.get("isPrimary").getAsBoolean()) {
+                  if (hasPrimaryLoginMethod) {
+                      errors.add("No two loginMethods can have isPrimary as true.");
+                  }
+                  hasPrimaryLoginMethod = true;
+              }
+          }
+
+          String recipeId = parseAndValidateField(jsonLoginMethodObj, "recipeId", ValueType.STRING, true, String.class, errors, " for a loginMethod.");
+          String tenantId = parseAndValidateField(jsonLoginMethodObj, "tenantId", ValueType.STRING, false, String.class, errors, " for a loginMethod.");
```

Comment:
there are other changes done on tenantId as well.. like it's made lowecase, we also check for certain strings not being present in this etc.

Reply:
We are not creating the tenant here, I don't think we need to normalise the input before checking if its valid or not.